api/src/main/java/org/openmrs/module/queue/tasks/AutoCloseVisitQueueEntryTask.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.queue.tasks;
+
+import java.util.Collection;
+import java.util.Date;
+
+import lombok.extern.slf4j.Slf4j;
+import org.openmrs.Visit;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.queue.api.QueueEntryService;
+import org.openmrs.module.queue.api.VisitQueueEntryService;
+import org.openmrs.module.queue.model.QueueEntry;
+import org.openmrs.module.queue.model.VisitQueueEntry;
+
+/**
+ * This iterates over all active VisitQueueEntries If the Visit associated with any of these has
+ * ended (non-null stopDatetime), then the QueueEntry associated with it is also ended at the same
+ * datetime as the Visit was stopped.
+ */
+@Slf4j
+public class AutoCloseVisitQueueEntryTask implements Runnable {
+	
+	private static boolean currentlyExecuting = false;
+	
+	@Override
```

Comment:
These are the crux of the changes.  Whereas the previous approach simply closed every active queue entry every 24 hours, this approach removes that for one that is more limited in it's scope - only automatically close active queue entries if they are associated with active visits.  Because the goal is to constantly check to see if there are any visits that are out of sync with their queue entries, this is executed every minute and set to ensure that if it happens to take a minute to execute (it shouldn't) that it doesn't run again in parallel.

Reply:
This will work well if all visits could be closed at the end of day or as appropriate. Our main challenge initially was about active visits that span several days before they are closed as they still have active queue entries. With proper patient check out, this approach will ensure near realtime queue clean up